### ISSUE TITLE
Replay of OpenFAST calculations takes a long time for coupled AMR-Wind/OpenFAST simulation re-start runs

### ISSUE THREAD

**Issue by ghost** [Exawind/amr-wind#466]
Currently, to restart a coupled AMR-Wind/OpenFAST simulation from a previous checkpoint, a user specifies a non-zero start time for OpenFAST, e.g. `Actuator.TurbineFastLine.openfast_start_time=60.0` for a prior 60-second simulation. OpenFAST will first “replay” the prior history, after which AMR-Wind and OpenFAST advance in time beyond 60s.

This replay is done for one turbine after another in a serial mode. When there are a lot of turbines, when the prior history is long, and when the timestep is small, this process itself will take an incredibly long time. As an example, a recent 20-turbine simulation took roughly 1hr **per turbine** to replay a 2-min history at 0.00625s timestep.

This poses as an operational bottleneck for large simulations which may require multiple restarts to complete. Could there be a better/lighter way to do such restarts? @sayerhs @shashankNREL Thank you!

**Comment by sayerhs** [comment]
The serialized execution of turbines is occurring because of the MPI `Bcast` call during initialization here (line 152): https://github.com/Exawind/amr-wind/blob/650998941d68cca0e236081be0dd13aa43d33975/amr-wind/wind_energy/actuator/turbine/fast/turbine_fast_ops.H#L140-L153

Even though the FAST instances are allocated to different MPI ranks they don't proceed in parallel as all the MPI ranks are waiting on the Bcast call for each turbine. We don't need the `replay` action to complete before performing the `Bcast` call, so one approach would be to remove the `replay_turbine` call from https://github.com/Exawind/amr-wind/blob/650998941d68cca0e236081be0dd13aa43d33975/amr-wind/wind_energy/actuator/turbine/fast/FastIface.cpp#L177-L181

and then call the `replay` action after the initialization data has been performed in `InitDataOp` here: https://github.com/Exawind/amr-wind/blob/650998941d68cca0e236081be0dd13aa43d33975/amr-wind/wind_energy/actuator/turbine/fast/turbine_fast_ops.H#L175-L176

**Comment by sayerhs** [comment]
@jl477 Can you also please confirm that the root process for each turbine is a different MPI rank and that they're not all being allocated to the same MPI rank?

**Comment by jl477** [comment]
Hi @sayerhs, thank you very much for responding! I'll try your comment above and report back if I manage to get it going.

I don't know how to find out whether "the root process for each turbine is a different MPI rank". Could you tell us @tchatte3 how? I did a quick search in the log file for a 20-turbine simulation, but I'm not sure if it gives the information that you asked for. 
```
10km_10m_M3.1_6_10min_notRefined$grep 'MPI' ALM20_MS_10km_10m.log.1214831
  MPI              :: ON    (Num. ranks = 522)
CUDA initialized with 1 GPU per MPI rank; 522 GPU(s) used in total
MPI initialized with 522 MPI processes
MPI initialized with thread support level 0
```

**Comment by sayerhs** [comment]
@jl477 @tchatte3

You can add a print statement before L177 https://github.com/Exawind/amr-wind/blob/650998941d68cca0e236081be0dd13aa43d33975/amr-wind/wind_energy/actuator/turbine/fast/turbine_fast_ops.H#L177 

```C++
if (info.is_root_proc) {
    std::cout << "Turbine ID: " << info.id << " " << info.label << " " << info.root_proc << std::endl;
}
```

This should print the integer ID for the turbine, the user-assigned turbine label (name) and the process that is handling the FAST instance for that particular turbine.

**Comment by jl477** [comment]
Thank you @sayerhs! The print out below is for a 20-tubine simulation using 6 Summit nodes or 36 MPI ranks:
```
run_mpi1$grep 'Turbine ID:' out.log
Turbine ID: 0 WTG01 29
Turbine ID: 1 WTG02 2
Turbine ID: 2 WTG03 3
Turbine ID: 3 WTG04 0
Turbine ID: 4 WTG05 5
Turbine ID: 5 WTG06 1
Turbine ID: 6 WTG07 6
Turbine ID: 7 WTG08 4
Turbine ID: 8 WTG09 10
Turbine ID: 9 WTG10 7
Turbine ID: 10 WTG11 8
Turbine ID: 11 WTG12 15
Turbine ID: 12 WTG13 11
Turbine ID: 13 WTG14 34
Turbine ID: 14 WTG15 16
Turbine ID: 15 WTG16 33
Turbine ID: 16 WTG17 21
Turbine ID: 17 WTG18 9
Turbine ID: 18 WTG19 12
Turbine ID: 19 WTG20 13
```

As to the approach you suggested above, I:
- commented out L179 `fast_replay_turbine(fi);` in `FastIface.cpp`
- added a line `fast_replay_turbine(data);` after L176 in `turbine_fast_ops.H`
- copied and pasted L239-283 which defines `fast_replay_turbine` from `FastIface.cpp` into `turbine_fast_ops.H`

I have no C programming experience and this is perhaps too naive an implementation of what you suggested, so I couldn't get the code to compile. If you could elaborate more on what to do exactly I can give it another try. Thanks!

**Comment by lawrenceccheung** [comment]
The changes suggested above have been implemented in my branch at https://github.com/lawrenceccheung/amr-wind/tree/fix_fast_replay.  However, I haven't tested it, so we'll need to make sure that it behaves properly.

Lawrence

**Comment by marchdf** [comment]
Hi all, what's the status of this issue? Was it resolved or still worth keeping open?

**Comment by lawrenceccheung** [comment]
The original issue that this refers to has been solved by Ganesh and others, so we can close it.  Probably worth mentioning that openfast restarts on Frontier still need attention in the main code, but that's a separate issue =).

Lawrence

**Comment by marchdf** [comment]
and that's last thing @lawrenceccheung mentioned is being tracked here: #984
